File: Leetcode_EX_3E8/1629. Slowest Key/slowest_key.cpp
// Leetcode 1629. Slowest Key
// https://leetcode.com/problems/slowest-key/
// time complexity: O(n)
// space complexity: O(1)

class Solution 
{
public:
    char slowestKey(vector<int>& releaseTimes, string keysPressed) 
    {
        char result = keysPressed[0];
        int max_gap = releaseTimes[0];
        for (int i = 1; i < releaseTimes.size(); i++)
        {
            int gap = releaseTimes[i] - releaseTimes[i - 1];
            if (gap > max_gap)
            {
                max_gap = gap;
                result = keysPressed[i];
            }
            else if (gap == max_gap)
            {
                result = max(result, keysPressed[i]);
            }
        }
        
        return result;
    }
};

File: Leetcode_EX_3E8/1629. Slowest Key/slowest_key.rs
impl Solution {
    pub fn slowest_key(release_times: Vec<i32>, keys_pressed: String) -> char {
        let mut max_gap = release_times[0];
        let strs: Vec<char> = keys_pressed.chars().collect();
        let mut result = strs[0];
        
        let mut i = 1;
        while i < release_times.len() {
            let gap = release_times[i] - release_times[i - 1];
            if gap > max_gap {
                max_gap = gap;
                result = strs[i];
            } else if gap == max_gap {
                if strs[i] > result {
                    result = strs[i];
                }
            }
            i += 1;
        }
        
        return result;
    }
}
